internal/yml/promote/merge_test.go

```diff
@@ -0,0 +1,174 @@
+package promote_test
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/nestoca/joy/internal/yml"
+	"github.com/nestoca/joy/internal/yml/promote"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v3"
+)
+
+func TestYmlMerge(t *testing.T) {
+	cases := []struct {
+		Name     string
+		Src      string
+		Dst      string
+		Joy      string
+		Proposal string
+	}{
+		{
+			Name:     "conflicting types",
+			Src:      "{key: 1}",
+			Dst:      "{key: hello}",
+			Joy:      "{key: 1}",
+			Proposal: "{key: 1}",
+		},
+		{
+			Name:     "conflicting types locked dst",
+			Src:      "{key: 1}",
+			Dst:      "{key: !lock hello}",
+			Joy:      "{key: !lock hello}",
+			Proposal: "{key: !lock hello}",
+		},
+		{
+			Name:     "conflicting types inner locked dst",
+			Src:      "{key: 1}",
+			Dst:      "{key: [!lock hello]}",
+			Joy:      "{key: 1}",
+			Proposal: "{key: 1}",
+		},
+		{
+			Name:     "conflicting types locked src",
+			Src:      "{key: !lock 1}",
+			Dst:      "{key: hello}",
+			Joy:      "{key: hello}",
+			Proposal: "{key: hello}",
+		},
+		{
+			Name:     "seq",
+			Src:      "[1, 2, 3]",
+			Dst:      "[]",
+			Joy:      "[1, 2, 3]",
+			Proposal: "[1, 2, 3]",
+		},
+		{
+			Name:     "locked seq",
+			Src:      "!lock [1, 2, 3]",
+			Dst:      "[]",
+			Joy:      "!lock [1, 2, 3]",
+			Proposal: "!lock [!!int TODO, !!int TODO, !!int TODO]",
+		},
+		{
+			Name:     "seq with dst longer",
+			Src:      "[1, 2, 3]",
+			Dst:      "[4, 5, 6, 7, 8]",
+			Joy:      "[1, 2, 3]",
+			Proposal: "[1, 2, 3]",
+		},
+		{
+			Name:     "seq with dst longer and lock",
+			Src:      "[1, 2, 3]",
+			Dst:      "[4, 5, 6, 7, !lock 8]",
+			Joy:      "[1, 2, 3]",
+			Proposal: "[1, 2, 3, !lock 8]",
+		},
+		{
+			Name:     "seq with dst inner lock",
+			Src:      "[1, 2, 3]",
+			Dst:      "[4, !lock 5, 6, 7, !lock 8]",
+			Joy:      "[1, 2, 3]",
+			Proposal: "[1, !lock 5, 3, !lock 8]",
```

Comment:
@ayahajar raises a good point. I don't think we should replace based on the index since that behaviour would be, at least to me, quite confusing.

We could either do: 

1. Insert the locked value where it makes sense: `[1, !lock 5, 2, 3, !lock8]`
1. Insert the locked value at the end: `[1, 2, 3, !lock 5, !lock8]`

I think option 2 is the least surprising, but yeah, we can do a lot of bike-shedding about this whole subject. 😅

If someone cares about the order and exact content, they should lock the whole array.

`!lock[4, 5, 6, 7]`

Reply:
I think in practice, most use-cases will end up locking the entire value. I tend to be pro index merging, because I imagine that if partial locking is important it would generally be like so:

```
key:
  - !lock env-specific
  - alias
  - alias
```

Where we can understand that the first element is meant to be locked across environments.

Also if we push locks to the end of the sequence, we can't correlate items that were pushed from an original source. Imagine this:

`[!lock a]` gets promoted to an env and becomes: `[!lock TODO]` the user fixes up the value `[!lock b]`.
Now if we promote again how do we know that the two values are correlated? Should it become: `[!lock TODO, !lock b]`?

Basically I think that in the same way that nodes are merged in maps by their keys, nodes should be merged in sequences by their index, otherwise a lot of book keeping needs to be done for a very niche scenario. Also when explaining how sequence merging works, its much easier to say "by index" than to explain when locks get moved to the back or are consolidated with source.

Reply:
Also, another point, if ordering does matter, then moving locked values to the back is breaking.

Reply:
> Also, another point, if ordering does matter, then moving locked values to the back is breaking.

Yes. 🙂 

>> If someone cares about the order and exact content, they should lock the whole array.
>> !lock[4, 5, 6, 7]

At this point, I am wondering if @alexstojda @silphid or @riclima have any opinions about ordering.

Reply:
I slept on it and here's my perspective: to me it makes more sense locking an index than locking a value (AKA @davidmdm suggestion).

> and imply the users have checked the src and dst values ahead of time.

Makes sense in general, but in our use case I think they should "know" what the src and dst values are ahead of time just by looking at the src because the blessed pattern we want to incentivize is to have environments be as similar to each other as possible. Having them vary so much that there are complex scenarios that you need to carefully double check before promoting should be considered an anti-pattern and we should steer users away from it. Ideally promoting a sequence from one environment to another with a lock would mean "give me this exact sequence, but I would like to change the second item of the list."

> Also, another point, if ordering does matter, then moving locked values to the back is breaking.

This becomes important because of this point. There are things that we use in the cluster that ordering does matter. Take for instance this somewhat contrived example derived from a deployment in our cluster:

```yaml
readinessProbe:
  exec:
    command:
       - !lock fee
       - -fi
       - !lock fo
       - -fum
```

List ordering does matter, so if whatever reason we want to change the command `fee` for something else on a different environment and/or (more likely) the value for a flag `-fi`, they're gonna be bumped to the back of the line, requiring a manual reordering on every promote instead of just filling in the `TODO` values.

> We could do:
>
> Insert the locked value where it makes sense: `[1, !lock 5, 2, 3, !lock8]`
> Insert the locked value at the end: `[1, 2, 3, !lock 5, !lock8]`

For this argument I'm not entirely sure my understanding is correct so please correct me if I'm wrong:

If order doesn't matter, then correlation becomes important: say I have `[!lock fee, fi, fo, fum]`. I do a first promote of this sequence on Tuesday from `a` to `b`. Now `b` has: `[fi, fo, fum, TODO]`. I replace the `TODO` value for `prod`, check that all the items I want are there, and move on. Notably, because order doesn't matter, I don't bother to reorder the list, so we gather a slight configuration drift. Then on Wednesday I do some more work on other unrelated parts of `a` and promote it to `b`. Surprise, now I have `[fi, fo fum, prod, TODO]`, because we can't correlate back that we already handled that item. Now every time I promote from  have to keep deleting that `TODO`. Maybe one day I forget and that causes an issue, etc.

I think index locking instead of value locking is simpler, easier to explain to others, and will require less cleaning up during promotes, so I vote for that.

Reply:
Just like @riclima, I also slept on it. I think that the alternative to index based merging is not feasible (or at least significantly more difficult) because we want our promotions to be `idempotent`.

If I promote qa to prod, there may be some changes and that all well and good. If I promote qa to prod again without any changes to prod, we should not create more changes since we are already meant to be synced. IE multiple promotes of the same env should not create more than one change. Index based merges guarantees that.

Reply:
Alright, let's do this then.